File: cpp/src/bitcoin/coinbase.cpp
#include "bitcoin/coinbase.hpp"

#include <stdexcept>

#include "bitcoin/serialize.hpp"
#include "util/endian.hpp"
#include "util/varint.hpp"

namespace erikslund::bitcoin {

namespace {

void append_output(Bytes& bytes, const CoinbaseOutput& output) {
    util::append_le64(bytes, output.value);
    append(bytes, util::encode_varint(output.script.size()));
    append(bytes, output.script);
}

} // namespace

Bytes build_coinbase1(ByteView script_sig_prefix, size_t extranonce_total, ByteView tag,
                      uint32_t version) {
    const size_t scriptsig_length = script_sig_prefix.size() + extranonce_total + tag.size();
    if (scriptsig_length > kMaxScriptSig)
        throw std::invalid_argument("coinbase scriptSig too long (> 100 bytes)");

    Bytes out;
    util::append_le32(out, version);
    out.push_back(0x01);

    Bytes prevout_null(36, 0x00);
    prevout_null[32] = prevout_null[33] = prevout_null[34] = prevout_null[35] = 0xff;
    append(out, prevout_null);

    append(out, util::encode_varint(scriptsig_length));
    append(out, script_sig_prefix);
    return out;
}

Bytes build_coinbase1(int64_t height, size_t extranonce_total, ByteView tag, uint32_t version) {
    return build_coinbase1(serialize_height(height), extranonce_total, tag, version);
}

Bytes build_coinbase2(const std::vector<CoinbaseOutput>& outputs,
                      const std::vector<CoinbaseOutput>& required_outputs, ByteView tag,
                      uint32_t sequence, uint32_t lock_time) {
    Bytes out;
    append(out, tag);
    util::append_le32(out, sequence);

    const size_t output_count = outputs.size() + required_outputs.size();
    append(out, util::encode_varint(output_count));

    for (const auto& output : outputs)
        append_output(out, output);
    for (const auto& output : required_outputs)
        append_output(out, output);

    util::append_le32(out, lock_time);
    return out;
}

Bytes legacy_to_witness(ByteView legacy_coinbase, ByteView witness_item) {
    if (legacy_coinbase.size() < 8)
        throw std::invalid_argument("legacy_to_witness: coinbase too short");
    if (witness_item.size() != 32)
        throw std::invalid_argument("legacy_to_witness: witness item must be 32 bytes");

    const ByteView version = legacy_coinbase.subspan(0, 4);
    const ByteView body = legacy_coinbase.subspan(4, legacy_coinbase.size() - 8);
    const ByteView locktime = legacy_coinbase.subspan(legacy_coinbase.size() - 4, 4);

    Bytes out;
    append(out, version);
    out.push_back(0x00);
    out.push_back(0x01);
    append(out, body);

    append(out, util::encode_varint(1));
    append(out, util::encode_varint(witness_item.size()));
    append(out, witness_item);

    append(out, locktime);
    return out;
}

Bytes legacy_to_witness(ByteView legacy_coinbase) {
    return legacy_to_witness(legacy_coinbase, Bytes(32, 0));
}

} // namespace erikslund::bitcoin

File: python/erikslund_pool/coinbase.py
"""Coinbase assembly and the coinbase1/coinbase2 Stratum split.

    full_legacy_coinbase = coinbase1 || extranonce1 || extranonce2 || coinbase2

coinbase1 is shared; coinbase2 carries the miner's own payout. The merkle-root
txid uses this legacy form; the submitted block re-serializes it as segwit.
"""

from erikslund_pool.constants import _LOCKTIME
from erikslund_pool.constants import _PREVOUT_NULL
from erikslund_pool.constants import _SEQUENCE
from erikslund_pool.constants import WITNESS_RESERVED_VALUE
from erikslund_pool.exceptions import WorkError
from erikslund_pool.util import ser_uint32
from erikslund_pool.util import ser_uint64
from erikslund_pool.util import ser_varint
from erikslund_pool.util import serialize_height


def build_coinbase1(height: int, extranonce_total: int, tag: bytes, version: int = 1) -> bytes:
    """coinbase1 prefix, ending where extranonce1 begins (scriptSig = height||enonce||tag)."""
    height_push = serialize_height(height)
    scriptsig_len = len(height_push) + extranonce_total + len(tag)
    if scriptsig_len > 100:
        raise WorkError(f"coinbase scriptSig too long ({scriptsig_len} > 100)")
    return ser_uint32(version) + b"\x01" + _PREVOUT_NULL + ser_varint(scriptsig_len) + height_push


def build_coinbase2(payout_script: bytes, value: int, witness_commitment_script: bytes | None,
                 tag: bytes, donation_script: bytes = b"", donation_percent: float = 0.0) -> bytes:
    """Per-miner coinbase2: scriptSig tail, sequence, outputs, locktime.

    With a donation, that percent of the reward goes to donation_script.
    """
    if donation_percent > 0.0 and donation_script:
        donation_amount = int(value * donation_percent / 100.0)
        outputs = [(value - donation_amount, payout_script), (donation_amount, donation_script)]
    else:
        outputs = [(value, payout_script)]
    if witness_commitment_script is not None:
        outputs.append((0, witness_commitment_script))
    coinbase2 = bytearray(tag)
    coinbase2 += _SEQUENCE
    coinbase2 += ser_varint(len(outputs))
    for amount, script in outputs:
        coinbase2 += ser_uint64(amount) + ser_varint(len(script)) + script
    coinbase2 += _LOCKTIME
    return bytes(coinbase2)


def legacy_to_witness(legacy_coinbase: bytes) -> bytes:
    """Re-serialize a legacy coinbase as segwit: marker+flag after nVersion, the
    input's 32-byte reserved-value witness before nLockTime."""
    version, body, locktime = legacy_coinbase[:4], legacy_coinbase[4:-4], legacy_coinbase[-4:]
    witness = ser_varint(1) + ser_varint(len(WITNESS_RESERVED_VALUE)) + WITNESS_RESERVED_VALUE
    return version + b"\x00\x01" + body + witness + locktime
